src/broker/boot_pmi.c

```diff
@@ -153,6 +153,28 @@ static int format_bind_uri (char *buf, int bufsz, attr_t *attrs, int rank)
     return -1;
 }
 
+static int set_hostlist_attr (attr_t *attrs, struct hostlist *hl)
+{
+    const char *value;
+    char *s;
+    int rc = -1;
+
+    /*  Allow hostlist attribute to be set on command line for testing.
+     *  The value must be re-added if so, so that the IMMUTABLE flag can
+     *   be set so that the attribute is properly cached.
+     */
+    if (attr_get (attrs, "hostlist", &value, NULL) == 0) {
+        s = strdup (value);
+        (void) attr_delete (attrs, "hostlist", true);
+    }
+    else
+        s = hostlist_encode (hl);
+    if (s && attr_add (attrs, "hostlist", s, FLUX_ATTRFLAG_IMMUTABLE) == 0)
+        rc = 0;
+    free (s);
```

Comment:
`free()` potentially clobbers errno, which needs to be set so `log_err()` prints something sensible on failure.